onetl/base/base_file_limit.py

```diff
@@ -15,30 +15,101 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
+from typing import TypeVar
 
 from onetl.base.path_protocol import PathProtocol
 
+Self = TypeVar("Self", bound="BaseFileLimit")
+
 
 class BaseFileLimit(ABC):
     """
-    Base file limit class
+    Base file limit interface.
+
+    Limits used by several onETL components, including :ref:`file-downloader` and :ref:`file-mover`,
+    to determine if internal loop should be stopped.
+
+    Unlike file filters, limits have internal state which can be updated or reset.
     """
 
     @abstractmethod
-    def reset(self):
+    def reset(self: Self) -> Self:
```

Comment:
if you start adding `self` typing to methods, then you probably need to add it to all methods below

Reply:
This annotation is needed only to show that method returns an object of the same class. Unlike plain `-> BaseFileFilter`, this also properly handles class inheritance.
mypy could detect this if `return self` was added into the method, but it is abstract, so this is the only way to properly type that.

Other methods return only boolean, so there is no actual need to add everywhere `self: Self`.